openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,489 @@
+            serialized_type = o['oml-python:serialized_object']
+            value = o['value']
+            if serialized_type == 'type':
+                rval = deserialize_type(value, **kwargs)
+            elif serialized_type == 'rv_frozen':
+                rval = deserialize_rv_frozen(value, **kwargs)
+            elif serialized_type == 'function':
+                rval = deserialize_function(value, **kwargs)
+            elif serialized_type == 'component_reference':
+                value = flow_to_sklearn(value)
+                step_name = value['step_name']
+                key = value['key']
+                component = flow_to_sklearn(kwargs['components'][key])
+                if step_name is None:
+                    rval = component
+                else:
+                    rval = (step_name, component)
+            elif serialized_type == 'cv_object':
+                rval = _deserialize_cross_validator(value, **kwargs)
+            else:
+                raise ValueError('Cannot flow_to_sklearn %s' % serialized_type)
+
+        else:
+            # Regular dictionary
+            rval = OrderedDict((flow_to_sklearn(key, **kwargs),
+                                flow_to_sklearn(value, **kwargs))
+                               for key, value in o.items())
+    elif isinstance(o, (list, tuple)):
+        rval = [flow_to_sklearn(element, **kwargs) for element in o]
+        if isinstance(o, tuple):
+            rval = tuple(rval)
+    elif isinstance(o, (bool, int, float)):
+        rval = o
+    elif isinstance(o, six.string_types):
+        rval = o
+    elif o is None:
+        rval = None
+    elif isinstance(o, OpenMLFlow):
+        rval = _deserialize_model(o, **kwargs)
+    else:
+        raise TypeError(o)
+    assert o is None or rval is not None
+
+    return rval
+
+
+def _serialize_model(model):
+    """Create an OpenMLFlow.
+
+    Calls `sklearn_to_flow` recursively to properly serialize the
+    parameters to strings and the components (other models) to OpenMLFlows.
+
+    Parameters
+    ----------
+    model : sklearn estimator
+
+    Returns
+    -------
+    OpenMLFlow
+
+    """
+    # stores all entities that should become subcomponents
+    sub_components = OrderedDict()
+    # stores the keys of all subcomponents that should become
+    sub_components_explicit = set()
+    parameters = OrderedDict()
+    parameters_meta_info = OrderedDict()
+
+    model_parameters = model.get_params(deep=False)
+
+    for k, v in sorted(model_parameters.items(), key=lambda t: t[0]):
+        rval = sklearn_to_flow(v)
+
+        if (isinstance(rval, (list, tuple)) and len(rval) > 0 and
+                isinstance(rval[0], (list, tuple)) and
+                [type(rval[0]) == type(rval[i]) for i in range(len(rval))]):
+
+            # Steps in a pipeline or feature union
+            parameter_value = list()
+            for sub_component_tuple in rval:
+                identifier, sub_component = sub_component_tuple
```

Comment:
I extracted the whole loop to get model information into its own function. This makes the `_serialize_model()` easier to read. Also, I added information on how to further restructure the code in case one has to touch it again.